### PR TITLE
Resolve Compare FFT implementations #62

### PR DESCRIPTION
**Pull request by einar-taiko**
This resolves `https://github.com/taikoxyz/zkevm-circuits/issues/62`

- [x] Fast-forward `main` to [`main`](https://github.com/privacy-scaling-explorations/halo2) from PSE.

- [x] Rebase Brecht's FFT fork on `main`
- [x] Rebase Scroll's FFT fork on `main`
- [x] Factor out redundancies between forks
- [x] Reuse tests from branches

- [x] Verify `cargo test`

- [x] Please `cargo check --all-targets`
- [x] Please `cargo clippy --all-targets`

### CODE REVIEW

**Review by Brechtpd — COMMENTED** [review]
Diff much improved!

File: halo2_proofs/src/arithmetic.rs [hunk 1]
```diff
@@ -496,28 +428,51 @@ pub(crate) fn powers<F: Field>(base: F) -> impl Iterator<Item = F> {
...
+    for _ in 0..l {
+        r = (r << 1) | (n & 1);
+        n >>= 1;
+    }
+    r
+}
 
 #[cfg(test)]
-use crate::halo2curves::pasta::Fp;
+mod tests {
```
[the remark this answers is not in the document]
Comment (einar-taiko) [reply halo2_proofs/src/arithmetic.rs]: There are reasons, but not related to this PR anyway.  Reverting.

File: halo2_proofs/src/plonk/evaluation.rs [hunk 2]
```diff
@@ -303,12 +303,13 @@ impl<C: CurveAffine> Evaluator<C> {
...
 
         // Calculate the advice and instance cosets
+        let start = start_measure("cosets", false);
         let advice: Vec<Vec<Polynomial<C::Scalar, ExtendedLagrangeCoeff>>> = advice_polys
             .iter()
             .map(|advice_polys| {
                 advice_polys
                     .iter()
-                    .map(|poly| domain.coeff_to_extended(poly.clone()))
+                    .map(|poly| domain.coeff_to_extended(poly.to_owned()))
```
[the remark this answers is not in the document]
Comment (einar-taiko) [reply halo2_proofs/src/plonk/evaluation.rs]: Same.

File: halo2_proofs/src/poly.rs [hunk 3]
```diff
@@ -147,7 +147,7 @@ impl<F, B> Polynomial<F, B> {
 }
 
 impl<F: SerdePrimeField, B> Polynomial<F, B> {
-    /// Reads polynomial from buffer using `SerdePrimeField::read`.  
+    /// Reads polynomial from buffer using `SerdePrimeField::read`.
```
[the remark this answers is not in the document]
Comment (einar-taiko) [reply halo2_proofs/src/poly.rs]: Let's keep it minimal.  I found that VSCode has a setting to only reformat new modifications, so it's extremely manageable.

File: halo2_proofs/src/poly/domain.rs [hunk 4]
```diff
@@ -53,6 +67,7 @@ impl<F: WithSmallOrderMulGroup<3>> EvaluationDomain<F> {
         while (1 << extended_k) < (n * quotient_poly_degree) {
             extended_k += 1;
         }
+        println!("k: {}, extended_k: {}", k, extended_k);
```
[the remark this answers is not in the document]
Comment (einar-taiko) [reply halo2_proofs/src/poly/domain.rs]: Done.

**Comment by einar-taiko** [comment]
@Brechtpd I have reverted as much as possible to produce a minimal diff including the things you mentioned.

I have also chosen `recursive_fft` as the default now.

Please have another look at the diff and see if there are further changes you would like.

File: halo2_proofs/src/poly/domain.rs [hunk 5]
```diff
@@ -360,6 +385,27 @@ impl<F: WithSmallOrderMulGroup<3>> EvaluationDomain<F> {
...
+            let fft_data = if a.len() == self.fft_data.n {
+                &self.fft_data
+            } else {
+                &self.extended_fft_data
+            };
+            recursive::fft(a, omega, log_n, fft_data, inverse);
+        } else {
+            // TODO
+            best_fft(a, omega, log_n, &FFTData::default(), false);
+        }
```
Comment (Brechtpd) [inline halo2_proofs/src/poly/domain.rs:400]:
```suggestion
        let fft_data = if a.len() == self.fft_data.n {
            &self.fft_data
        } else {
            &self.extended_fft_data
        };
        best_fft(a, omega, log_n, fft_data, false);
```

Should use the selected FFT everywhere.  `get_fft_mode` shouldn't be used anymore as your `var("FFT")` env variable is now used instead.

File: halo2_proofs/src/poly/domain.rs [hunk 6]
```diff
@@ -245,7 +264,13 @@ impl<F: WithSmallOrderMulGroup<3>> EvaluationDomain<F> {
 
         self.distribute_powers_zeta(&mut a.values, true);
         a.values.resize(self.extended_len(), F::ZERO);
-        best_fft(&mut a.values, self.extended_omega, self.extended_k);
+        best_fft(
+            &mut a.values,
+            self.extended_omega,
+            self.extended_k,
+            &FFTData::default(),
```
Comment (Brechtpd) [inline halo2_proofs/src/poly/domain.rs:271]:
```suggestion
            &self.extended_fft_data,
```

Don't think this would otherwise work correctly with the recursive FFT?

File: halo2_proofs/src/fft.rs [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+/// Read Environment Variable `DEGREE`
+#[allow(dead_code)]
+pub fn get_degree() -> usize {
+    var("DEGREE")
+        .unwrap_or_else(|_| "22".to_string())
+        .parse()
+        .expect("Cannot parse DEGREE env var as usize")
+}
```
Comment (Brechtpd) [inline halo2_proofs/src/fft.rs:29]: `get_fft_mode` and `get_degree` can simply be removed I think.

File: halo2_proofs/benches/fft.rs [hunk 8]
```diff
@@ -16,7 +17,7 @@ fn criterion_benchmark(c: &mut Criterion) {
             let mut a = (0..(1 << k)).map(|_| Fp::random(OsRng)).collect::<Vec<_>>();
             let omega = Fp::random(OsRng); // would be weird if this mattered
             b.iter(|| {
-                best_fft(&mut a, omega, k as u32);
+                best_fft(&mut a, omega, k as u32, &FFTData::default(), false);
```
Comment (Brechtpd) [inline halo2_proofs/benches/fft.rs:20]: I think this will fail when running this with the recursive FFT because `FFTData::default()` won't contain the valid data. However this is just a performance benchmark so ehhh if you can quickly generate the right data here (I think not hard) please do so, otherwise also ok to just leave it as is.

File: halo2_proofs/src/fft.rs (hunk 7 above)
Comment (einar-taiko) [reply halo2_proofs/src/fft.rs:29]: `get_fft_mode` is gone.
`get_degree` is still used in some tests, so I moved it to the FFT test module.

File: halo2_proofs/src/poly/domain.rs (hunk 6 above)
Comment (einar-taiko) [reply halo2_proofs/src/poly/domain.rs]: My mistake, thanks.  Fixed in latest.

File: halo2_proofs/benches/fft.rs (hunk 8 above)
Comment (einar-taiko) [reply halo2_proofs/benches/fft.rs]: Fixed in latest.

**Comment by einar-taiko** [comment]
@Brechtpd Please confirm which implementation you would like as default, keeping in mind that `recursive` is slightly faster for k=25 while it runs OOM for k=29.  Besides that, I'd say the PR is ready.

File: halo2_proofs/src/plonk/evaluation.rs [hunk 9]
```diff
@@ -442,14 +447,19 @@ impl<C: CurveAffine> Evaluator<C> {
...
                 // Polynomials required for this lookup.
                 // Calculated here so these only have to be kept in memory for the short time
                 // they are actually needed.
-                let product_coset = pk.vk.domain.coeff_to_extended(lookup.product_poly.clone());
-                let permuted_input_coset = pk
+                let product_coset = pk
+                    .vk
+                    .domain
+                    .coeff_to_extended(lookup.product_poly.clone());
+                let permuted_input_coset  = pk
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/evaluation.rs:462]: Unnecessary diff it looks like.

File: halo2_proofs/src/plonk/circuit.rs [hunk 10]
```diff
@@ -1,21 +1,23 @@
...
     ops::{Neg, Sub},
 };
 
+use super::{lookup, permutation, Assigned, Error};
+use crate::dev::metadata;
+use crate::{
+    circuit::{Layouter, Region, Value},
+    poly::Rotation,
+};
+
```
Comment (Brechtpd) [inline halo2_proofs/src/plonk/circuit.rs:20]: Unnecessary diff I think.

File: halo2_proofs/src/poly/domain.rs [hunk 11]
```diff
@@ -2,24 +2,31 @@
...
+
+/// TEMP
+pub static mut FFT_TOTAL_TIME: usize = 0;
 
 /// This structure contains precomputed constants and other details needed for
 /// performing operations on an evaluation domain of size $2^k$ and an extended
 /// domain of size $2^{k} * j$ with $j \neq 0$.
 #[derive(Clone, Debug)]
-pub struct EvaluationDomain<F: Field> {
+pub struct EvaluationDomain<F: ff::Field> {
```
Comment (Brechtpd) [inline halo2_proofs/src/poly/domain.rs:29]: This change necessary?

File: halo2_proofs/src/poly/domain.rs [hunk 12]
```diff
@@ -245,7 +264,17 @@ impl<F: WithSmallOrderMulGroup<3>> EvaluationDomain<F> {
...
+        let l = a.len();
+        let fft_data = self.get_fft_data(l);
+
+        best_fft(
+            &mut a.values,
+            self.extended_omega,
+            self.extended_k,
+            fft_data,
+            false,
+        );
```
Comment (Brechtpd) [inline halo2_proofs/src/poly/domain.rs:277]:
```suggestion
        best_fft(
            &mut a.values,
            self.extended_omega,
            self.extended_k,
            &self.extended_fft_data
            false,
        );
```

Always extended anyway.

File: halo2_proofs/src/poly/domain.rs [hunk 13]
```diff
@@ -474,17 +515,35 @@ impl<F: WithSmallOrderMulGroup<3>> EvaluationDomain<F> {
...
+    pub fn get_n(&self) -> u64 { self.n }
+
+    /// Get the private `fft_data`
+    pub fn get_fft_data(&self, l: usize) -> &FFTData<F> {
+            if l == self.fft_data.get_n() {
+                &self.fft_data
+            } else {
+                &self.extended_fft_data
+            }
+        }
```
Comment (Brechtpd) [inline halo2_proofs/src/poly/domain.rs:529]: Strange indentation?

**Review by Brechtpd — COMMENTED** [review]
Just some nitpicks mostly about the diff, otherwise LGTM!

File: halo2_proofs/src/plonk/evaluation.rs (hunk 9 above)
Comment (einar-taiko) [reply halo2_proofs/src/plonk/evaluation.rs:462]: Reverted.

File: halo2_proofs/src/poly/domain.rs (hunk 13 above)
Comment (einar-taiko) [reply halo2_proofs/src/poly/domain.rs:529]: Indeed.  It seems like my "Format Selection" feature in Vscode is broken.  Sorry for the inconvenience.

File: halo2_proofs/src/poly/domain.rs [hunk 14]
```diff
@@ -126,6 +138,7 @@ impl<F: WithSmallOrderMulGroup<3>> EvaluationDomain<F> {
             .chain(Some(&mut omega_inv))
             .batch_invert();
 
+
```
[the remark this answers is not in the document]
Comment (einar-taiko) [reply halo2_proofs/src/poly/domain.rs:141]: Done.

File: halo2_proofs/src/poly/domain.rs (hunk 12 above)
Comment (einar-taiko) [reply halo2_proofs/src/poly/domain.rs:277]: You are right.  I  made the field `pub`.

File: halo2_proofs/src/poly/domain.rs (hunk 11 above)
Comment (einar-taiko) [reply halo2_proofs/src/poly/domain.rs]: Not anymore,  good catch!  I had two conflicting `Field` types and prefixed them to stay sane.

File: halo2_proofs/src/poly/domain.rs [hunk 15]
```diff
@@ -474,17 +515,35 @@ impl<F: WithSmallOrderMulGroup<3>> EvaluationDomain<F> {
...
+                &self.extended_fft_data
+            }
+        }
 }
 
 /// Represents the minimal parameters that determine an `EvaluationDomain`.
 #[allow(dead_code)]
 #[derive(Debug)]
-pub struct PinnedEvaluationDomain<'a, F: Field> {
+pub struct PinnedEvaluationDomain<'a, F: ff::Field> {
```
[the remark this answers is not in the document]
Comment (einar-taiko) [reply halo2_proofs/src/poly/domain.rs]: Same as previous above.

**Comment by einar-taiko** [comment]
All suggestions implemented.

> @Brechtpd Please confirm which implementation you would like as default, keeping in mind that `recursive` is slightly faster for k=25 while it runs OOM for k=29. [...]

@Brechtpd Please let me know which you prefer as default before we merge.  It currently defaults to `baseline`.

File: halo2_proofs/src/plonk/evaluation.rs (hunk 9 above)
Comment (Brechtpd) [reply halo2_proofs/src/plonk/evaluation.rs:462]: I think this one is not yet resolved.

File: halo2_proofs/src/plonk/circuit.rs (hunk 10 above)
Comment (Brechtpd) [reply halo2_proofs/src/plonk/circuit.rs:20]: This one either I think.

**Comment by Brechtpd** [comment]
> @Brechtpd Please let me know which you prefer as default before we merge. It currently defaults to `baseline`.

Let's go with the parallel one.

**Comment by einar-taiko** [comment]
> Let's go with the parallel one.

Done.

**Comment by Brechtpd** [comment]
Still two diff changes left that I unresolved, see above. After that were done!

File: halo2_proofs/src/plonk/evaluation.rs (hunk 9 above)
Comment (einar-taiko) [reply halo2_proofs/src/plonk/evaluation.rs]: Not sure what happened here.  Should be reverted now.

File: halo2_proofs/src/plonk/circuit.rs (hunk 10 above)
Comment (einar-taiko) [reply halo2_proofs/src/plonk/circuit.rs]: Not sure what happened here.  Should be reverted now.

**Comment by einar-taiko** [comment]
> Still two diff changes left that I unresolved, see above. After that were done!

@Brechtpd Not sure what went wrong last night.   The last two should now be reverted in [5bc5ca5](https://github.com/taikoxyz/halo2/pull/4/commits/5bc5ca5b1025de46d096dbd6d5d48ddfb2893e7b).
